### PR TITLE
Add grains_refresh_pre_exec minion option

### LINKED ISSUE #61708: [FEATURE REQUEST] Allow grains to be refreshed before any operation
**Is your feature request related to a problem? Please describe.**
Environments using dynamic grains from external sources often have a need to populate grains information in another tool (such as a DCIM) and have those grains immediately take effect during the next execution on a minion. With the current workflow, grains can only be refreshed manually, on an interval (causing potentially unnecessary polling) or **after** a highstate (when we wanted the grains refreshed before the highstate).

**Describe the solution you'd like**
I'd like a way to refresh grains prior to any operation on a minion in order to ensure the minion has the most up-to-date information prior to execution.

**Describe alternatives you've considered**
Considered just using `grains_refresh_every` and also unsuccessfully tried a few different ways to refresh grains inside of Jinja and states.

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
This PR adds the `grains_refresh_pre_exec` minion option to gate grains refresh operations before every execution.

### What issues does this PR fix or reference?
Fixes: #61708

### Previous Behavior
See issue for details.

### New Behavior
The `grains_refresh_pre_exec` setting allows for a minion to check its grains prior to the execution of any operation to see if they have changed and, if so, to inform the master of the new grains.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by dmurphy18** [comment]
Asked in issue about saltutil.refresh_grains satisfying this need.

File: salt/minion.py [hunk 1]
```diff
@@ -1744,6 +1744,9 @@ def _handle_decoded_payload(self, data):
                 self.schedule.functions = self.functions
                 self.schedule.returners = self.returners
 
+        if self.opts["grains_refresh_pre_exec"]:
+            self.opts["grains"] = salt.loader.grains(self.opts, force_refresh=True)
```
Comment (dmurphy18) [inline salt/minion.py:1748]: Wonder if need to consider proxy minion usage and context as in similar code to load the grains in _load_modules, see https://github.com/saltstack/salt/blob/cbb22cfd2bb939864657df99f798443a48bde797/salt/minion.py#L1555-L1567

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2016-x64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2016-x64-py3-pytest

File: salt/minion.py (hunk 1 above)
Comment (nicholasmhughes) [reply salt/minion.py:1748]: Dunno. Tests passed. Not sure if there are any to check this aspect of proxy operation.

Comment (dmurphy18) [reply salt/minion.py:1748]: The proxy minion should be considered too,  could make a case for some embedded systems with dynamic characteristics which are accessed through a proxy that would benefit from that setting too

**Comment by nicholasmhughes** [comment]
re-run pr-windows-2016-x64-py3-pytest

**Review by dmurphy18 — APPROVED** [review]

**Review by whytewolf — APPROVED** [review]

**Pull request merged**